src/components/constraints/ConstraintForm.svelte

```diff
@@ -3,233 +3,372 @@
-          user,
-          initialPlans,
-          constraintDescription,
+      if (newConstraintId !== null) {
+        goto(
+          `${base}/constraints/edit/${newConstraintId}${
+            referenceModelId !== null ? `?${SearchParameters.MODEL_ID}=${referenceModelId}` : ''
+          }`,
         );
+      }
+    }
+  }
 
-        // Associate new tags with constraint
-        const newConstraintTags: ConstraintTagsInsertInput[] = (constraintTags || []).map(({ id: tag_id }) => ({
-          constraint_id: constraintId as number,
-          tag_id,
-        }));
-        await effects.createConstraintTags(newConstraintTags, user);
+  async function createNewConstraintDefinition() {
+    if (saveButtonEnabled && constraintMetadataId !== null) {
+      const definition = await effects.createConstraintDefinition(
+        constraintMetadataId,
+        constraintDefinitionCode,
+        constraintDefinitionTags.map(({ id }) => ({ tag_id: id })),
+        user,
+      );
 
-        // Disassociate old tags from constraint
-        const unusedTags = initialConstraintTags
-          .filter(tag => !constraintTags.find(t => tag.id === t.id))
-          .map(tag => tag.id);
-        await effects.deleteConstraintTags(unusedTags, user);
+      if (definition !== null) {
+        selectRevision(definition.revision);
+      }
+    }
+  }
+
+  function onSelectReferenceModel(event: CustomEvent<number | null>) {
+    const { detail } = event;
+    referenceModelId = detail;
+    dispatch('selectReferenceModel', detail);
+  }
 
-        savedConstraint = {
-          definition: constraintDefinition,
+  async function saveConstraint() {
+    if (constraintMetadataId) {
+      if (isMetadataModified) {
+        await saveConstraintMetadata();
+      }
+      if (isDefinitionDataModified && !isDefinitionModified) {
+        await saveConstraintDefinitionRevisionTags();
+      } else if (isDefinitionModified) {
+        await createNewConstraintDefinition();
+      }
+    } else {
+      await createConstraint();
+    }
+  }
+
+  async function saveConstraintMetadata() {
+    if (constraintMetadataId !== null) {
+      const tagsToUpdate: ConstraintMetadataTagsInsertInput[] = (constraintMetadataTags || []).map(
+        ({ id: tag_id }) => ({
+          constraint_id: constraintMetadataId as number,
+          tag_id,
+        }),
+      );
+      // Disassociate old tags from constraint
+      const tagIdsToDelete = savedConstraintMetadata.tags
+        .filter(({ tag }) => !constraintMetadataTags.find(t => tag.id === t.id))
+        .map(({ tag }) => tag.id);
+
+      await effects.updateConstraintMetadata(
+        constraintMetadataId,
+        {
           description: constraintDescription,
-          model_id: constraintModelId,
           name: constraintName,
-          plan_id: constraintPlanId,
-          tags: constraintTags.map(tag => ({ tag })),
-        };
-      }
+          owner: constraintOwner,
```

Comment:
If the owner does not exist the constraint fails to save – is there any way we could check existence of the user as a validation step of the owner field? Might be a future improvement. But otherwise you'd have to pop open the console logs to figure out why the constraint did not save.

Reply:
Is this happening (1) when the owner used to exist and has been deleted, (2) when there's a fresh DB and the user is creating a constraint without logging in (ie, due to having an unexpired but invalid `user` cookie), (3) when the user tries to pass ownership to a user that doesn't exist, or (4) some fourth set up?

Reply:
I was thinking of (3)

Reply:
Ohhh...I understand now. I think that might require further discussion that is outside the scope of this PR. I think that ideally there'd be potentially some searchable dropdown of sorts of all the users that have logged in to aerie.